secure/key-manager/src/tests.rs

```diff
@@ -251,58 +208,296 @@ impl LibraInterface for TestLibraInterface {
     }
 }
 
+impl LibraInterfaceWrapper for MockLibraInterface {
+    fn retrieve_discovery_set(&self) -> Result<DiscoverySet, Error> {
+        let account = account_config::discovery_set_address();
+        let account_state = self.retrieve_account_state(account)?;
+        Ok(account_state
+            .get_discovery_set_resource()?
+            .ok_or_else(|| Error::DataDoesNotExist("DiscoverySetResource".into()))?
+            .discovery_set()
+            .clone())
+    }
+
+    fn retrieve_libra_block_resource(&self) -> Result<LibraBlockResource, Error> {
+        let account = account_config::association_address();
+        let account_state = self.retrieve_account_state(account)?;
+        account_state
+            .get_libra_block_resource()?
+            .ok_or_else(|| Error::DataDoesNotExist("BlockMetadata".into()))
+    }
+
+    fn take_all_transactions(&self) -> Vec<Transaction> {
+        self.queued_transactions.borrow_mut().drain(0..).collect()
+    }
+}
+
+#[derive(Clone)]
+/// A JsonRpcLibraInterface wrapper implementation that stores a list of transactions submitted via
+/// the API and exposes additional methods for testing (by implementing the LibraInterfaceWrapper
+/// trait). This is required to better test the JsonRpcLibraInterface implementation.
+struct JsonRpcLibraWrapper {
+    submitted_transactions: Arc<RefCell<Vec<Transaction>>>,
+    libra_interface: JsonRpcLibraInterface,
+}
+
+impl LibraInterface for JsonRpcLibraWrapper {
+    fn libra_timestamp(&self) -> Result<u64, Error> {
+        self.libra_interface.libra_timestamp()
+    }
+
+    fn last_reconfiguration(&self) -> Result<u64, Error> {
+        self.libra_interface.last_reconfiguration()
+    }
+
+    fn retrieve_sequence_number(&self, account: AccountAddress) -> Result<u64, Error> {
+        self.libra_interface.retrieve_sequence_number(account)
+    }
+
+    fn submit_transaction(&self, transaction: Transaction) -> Result<(), Error> {
+        self.submitted_transactions
+            .borrow_mut()
+            .push(transaction.clone());
+        self.libra_interface.submit_transaction(transaction)
+    }
+
+    fn retrieve_validator_config(&self, account: AccountAddress) -> Result<ValidatorConfig, Error> {
+        self.libra_interface.retrieve_validator_config(account)
+    }
+
+    fn retrieve_validator_info(&self, account: AccountAddress) -> Result<ValidatorInfo, Error> {
+        self.libra_interface.retrieve_validator_info(account)
+    }
+}
+
+impl LibraInterfaceWrapper for JsonRpcLibraWrapper {
```

Comment:
Maybe LibraInterfaceTestHarness if we have to...

Reply:
Ooh, thanks! That's a better name!